src/Orc.DataAccess/Database/DbDataSourceProviders/Base/MaSqlDbDataSourceProviderBase.cs

```diff
@@ -0,0 +1,111 @@
+
+public abstract class MsSqlDbDataSourceProviderBase : IDbDataSourceProvider
+{
+    private static readonly ILog Log = LogManager.GetCurrentClassLogger();
+    
+    private const string MicrosoftSqlServerRegPath = @"SOFTWARE\Microsoft\Microsoft SQL Server";
+
+    private readonly IRegistryKeyService _registryKeyService;
+
+    protected MsSqlDbDataSourceProviderBase()
+        : this(ServiceLocator.Default.ResolveRequiredType<IRegistryKeyService>())
+    {
+    }
+
+    internal MsSqlDbDataSourceProviderBase(IRegistryKeyService registryKeyService)
+    {
+        ArgumentNullException.ThrowIfNull(registryKeyService);
+
+        _registryKeyService = registryKeyService;
+    }
+
+    protected abstract string ProviderName { get; }
+
+    public virtual IList<DbDataSource> GetDataSources()
+    {
+        var localServers = GetLocalSqlServerInstances().ToList();
+        var remoteServers = GetRemoteSqlServerInstances().ToList();
+
+        return localServers.Union(remoteServers)
+            .Distinct()
+            .OrderBy(x => x)
+            .Select(x => new DbDataSource(ProviderName, x))
+            .ToList();
+    }
+
+    private IEnumerable<string> GetLocalSqlServerInstances()
+    {
+        var machineName = Environment.MachineName;
+
+        var localSqlInstances32 = GetInstalledInstancesInRegistryView(RegistryView.Registry32);
+        var localSqlInstances64 = GetInstalledInstancesInRegistryView(RegistryView.Registry64);
+
+        return localSqlInstances32.Union(localSqlInstances64)
+            .Select(x => $"{machineName}" + (string.Equals(x, "MSSQLSERVER") ? string.Empty : $"\\{x}"));
+    }
+
+    private IEnumerable<string> GetInstalledInstancesInRegistryView(RegistryView registryView)
+    {
+        using var regView = _registryKeyService.OpenBaseKey(RegistryHive.LocalMachine, registryView);
+        using var sqlServerNode = regView.OpenSubKey(MicrosoftSqlServerRegPath);
+
+        return sqlServerNode?.GetValue("InstalledInstances") as IList<string> ?? Array.Empty<string>();
+    }
+
+    private IEnumerable<string> GetRemoteSqlServerInstances()
+    {
+        DbProviderFactory? sqlFactory;
+
+        try
+        {
+            sqlFactory = DbProviderFactories.GetFactory(ProviderName);
+        }
+        catch (Exception e)
+        {
+            Log.Error(e);
+
+            return Enumerable.Empty<string>();
+        }
+
+        if (!sqlFactory.CanCreateDataSourceEnumerator)
+        {
+            return Enumerable.Empty<string>();
+        }
+
+        var dataSourceEnumerator = sqlFactory.CreateDataSourceEnumerator();
+
+        using var dataTable = dataSourceEnumerator?.GetDataSources() ?? new DataTable();
+        var serversCount = dataTable.Rows.Count;
+        var servers = new string[serversCount];
+
+        for (var i = 0; i < serversCount; i++)
+        {
+            var name = dataTable.Rows[i]["ServerName"].ToString();
+            var instance = dataTable.Rows[i]["InstanceName"].ToString();
```

Comment:
fetch instance only after the string null check

Reply:
![image](https://github.com/WildGums/Orc.DataAccess/assets/5163834/439df500-0248-4d79-8528-f1b71a75b136)
null allowed